File: services/eval.go
package services

import (
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"slices"
	"strings"
	"time"

	"go.yaml.in/yaml/v3"

	"github.com/fmind/fkf/core"
)

const (
	evalSchemaVersion = 1
	evalDirectory     = "evals"
	evalQueriesFile   = "queries.yaml"
	maxEvalK          = 100
	maxEvalBudget     = int(core.MaxNarrativeBytes / 4)
)

// EvalSuite is the strict, versioned retrieval acceptance contract stored in a base.
type EvalSuite struct {
	FKF             int         `yaml:"fkf"`
	K               int         `yaml:"k"`
	Budget          *int        `yaml:"budget"`
	Delivery        string      `yaml:"delivery"`
	RecallThreshold float64     `yaml:"recall_threshold"`
	Queries         []EvalQuery `yaml:"queries"`
}

// EvalQuery names one question and the URIs its top-k answer must include or exclude.
type EvalQuery struct {
	Name          string   `yaml:"name"`
	Question      string   `yaml:"question"`
	Window        Window   `yaml:"window"`
	K             *int     `yaml:"k"`
	Budget        *int     `yaml:"budget"`
	Delivery      string   `yaml:"delivery"`
	ExpectEmpty   bool     `yaml:"expect_empty"`
	ExpectedURIs  []string `yaml:"expected_uris"`
	ForbiddenURIs []string `yaml:"forbidden_uris"`
}

// EvalExpectedRank records the final delivered rank for one expected URI. Rank zero means the
// requested delivery omitted it; a rank greater than k is present but still fails the top-k gate.
type EvalExpectedRank struct {
	URI  string `json:"uri"`
	Rank int    `json:"rank"`
}

// EvalQueryResult is one reproducible recall-at-k measurement.
type EvalQueryResult struct {
	Name            string             `json:"name"`
	Question        string             `json:"question"`
	Window          Window             `json:"window"`
	Budget          int                `json:"budget"`
	Delivery        string             `json:"delivery"`
	K               int                `json:"k"`
	ExpectEmpty     bool               `json:"expect_empty"`
	Recall          float64            `json:"recall"`
	Expected        int                `json:"expected"`
	FoundExpected   int                `json:"found_expected"`
	ExpectedRanks   []EvalExpectedRank `json:"expected_ranks"`
	MissingExpected []string           `json:"missing_expected,omitempty"`
	ForbiddenFound  []string           `json:"forbidden_found,omitempty"`
	DeliveredURIs   []string           `json:"delivered_uris"`
	DeliveredBytes  int                `json:"delivered_bytes"`
	DeliveredTokens int                `json:"delivered_tokens"`
	Index           LexicalIndexUse    `json:"index"`
	InputDigest     string             `json:"input_digest"`
	RankingVersion  int                `json:"ranking_version"`
	RecallThreshold float64            `json:"recall_threshold"`
	Passed          bool               `json:"passed"`
}

// EvalReport is the complete result for evals/queries.yaml. Evaluation never writes the base.
type EvalReport struct {
	Path            string            `json:"path"`
	K               int               `json:"k"`
	Budget          int               `json:"budget"`
	Delivery        string            `json:"delivery"`
	RecallThreshold float64           `json:"recall_threshold"`
	EvaluationTime  time.Time         `json:"evaluation_time"`
	Queries         []EvalQueryResult `json:"queries"`
	PassedQueries   int               `json:"passed_queries"`
	Failed          int               `json:"failed_queries"`
	Passed          bool              `json:"passed"`
}

// Evaluate runs every declared query against stored evidence and reports recall at k. It does
// not run a provider command, fetch a body, or update a derived cache.
func Evaluate(ctx context.Context, base *Base) (*EvalReport, error) {
	suite, relative, err := loadEvalSuite(ctx, base)
	if err != nil {
		return nil, err
	}
	evaluationTime := base.Now()
	report := &EvalReport{
		Path: relative, K: suite.K, Budget: *suite.Budget, Delivery: suite.Delivery, RecallThreshold: suite.RecallThreshold,
		EvaluationTime: evaluationTime,
		Queries:        make([]EvalQueryResult, 0, len(suite.Queries)),
	}
	for _, query := range suite.Queries {
		if err := checkContext(ctx); err != nil {
			return nil, err
		}
		result, err := evaluateQuery(ctx, base, suite, query, evaluationTime)
		if err != nil {
			return nil, fmt.Errorf("evaluate %s: %w", query.Name, err)
		}
		report.Queries = append(report.Queries, result)
		if result.Passed {
			report.PassedQueries++
		} else {
			report.Failed++
		}
	}
	report.Passed = report.Failed == 0
	return report, nil
}

func evaluateQuery(
	ctx context.Context,
	base *Base,
	suite *EvalSuite,
	query EvalQuery,
	evaluationTime time.Time,
) (EvalQueryResult, error) {
	budget := *suite.Budget
	if query.Budget != nil {
		budget = *query.Budget
	}
	k := suite.K
	if query.K != nil {
		k = *query.K
	}
	delivery := suite.Delivery
	if query.Delivery != "" {
		delivery = query.Delivery
	}
	pack, err := BuildContext(ctx, base, ContextRequest{
		Query: query.Question, Window: query.Window, Budget: budget, DeliveryFormat: delivery,
		evaluationTime: evaluationTime,
	})
	if err != nil {
		return EvalQueryResult{}, err
	}
	delivered := make([]string, 0, len(pack.Items))
	ranks := make(map[string]int, len(pack.Items))
	for index, item := range pack.Items {
		delivered = append(delivered, item.URI)
		ranks[item.URI] = index + 1
	}
	deliveredBytes := encodedBytes(pack)
	if delivery == ContextDeliveryText {
		deliveredBytes = len(RenderContextText(pack))
	}
	result := EvalQueryResult{
		Name: query.Name, Question: query.Question, Window: pack.Receipt.Window,
		Budget: budget, Delivery: delivery, K: k, ExpectEmpty: query.ExpectEmpty, Expected: len(query.ExpectedURIs),
		ExpectedRanks: make([]EvalExpectedRank, 0, len(query.ExpectedURIs)), DeliveredURIs: delivered,
		DeliveredBytes: deliveredBytes, DeliveredTokens: bytesToTokens(deliveredBytes),
		Index:       pack.Receipt.Index,
		InputDigest: pack.Receipt.InputDigest, RankingVersion: pack.Receipt.RankingVersion,
		RecallThreshold: suite.RecallThreshold,
	}
	for _, expected := range query.ExpectedURIs {
		rank := ranks[expected]
		result.ExpectedRanks = append(result.ExpectedRanks, EvalExpectedRank{URI: expected, Rank: rank})
		if rank > 0 && rank <= k {
			result.FoundExpected++
		} else {
			result.MissingExpected = append(result.MissingExpected, expected)
		}
	}
	for _, forbidden := range query.ForbiddenURIs {
		if slices.Contains(delivered, forbidden) {
			result.ForbiddenFound = append(result.ForbiddenFound, forbidden)
		}
	}
	if query.ExpectEmpty {
		result.Recall = 1
		if len(delivered) > 0 || pack.matchedButOmitted {
			result.Recall = 0
		}
		result.Passed = len(delivered) == 0 && !pack.matchedButOmitted
	} else {
		result.Recall = float64(result.FoundExpected) / float64(result.Expected)
		result.Passed = result.Recall >= suite.RecallThreshold && len(result.ForbiddenFound) == 0
	}
	return result, nil
}

func loadEvalSuite(ctx context.Context, base *Base) (*EvalSuite, string, error) {
	relative := filepath.ToSlash(filepath.Join(evalDirectory, evalQueriesFile))
	absolute := filepath.Join(base.Root(), filepath.FromSlash(relative))
	if err := core.ValidateWithinRoot(base.Root(), absolute); err != nil {
		return nil, relative, err
	}
	data, err := core.ReadFileLimitContext(ctx, absolute, core.MaxConfigBytes)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, relative, fmt.Errorf("%w: %s is missing; add the base's retrieval evaluation set", core.ErrConfig, relative)
		}
		return nil, relative, err
	}
	var suite EvalSuite
	decoder := yaml.NewDecoder(strings.NewReader(string(data)))
	decoder.KnownFields(true)
	if err := decoder.Decode(&suite); err != nil {
		return nil, relative, fmt.Errorf("%w: %s: %w", core.ErrConfig, relative, err)
	}
	var extra any
	if err := decoder.Decode(&extra); err == nil {
		return nil, relative, fmt.Errorf("%w: %s holds more than one YAML document", core.ErrConfig, relative)
	} else if !errors.Is(err, io.EOF) {
		return nil, relative, fmt.Errorf("%w: %s has invalid trailing YAML: %w", core.ErrConfig, relative, err)
	}
	if err := validateEvalSuite(&suite); err != nil {
		return nil, relative, fmt.Errorf("%w: %s: %w", core.ErrConfig, relative, err)
	}
	return &suite, relative, nil
}

func validateEvalSuite(suite *EvalSuite) error {
	if err := validateEvalSuiteDefaults(suite); err != nil {
		return err
	}
	names := make(map[string]struct{}, len(suite.Queries))
	for index := range suite.Queries {
		if err := validateEvalQuery(&suite.Queries[index], index, names); err != nil {
			return err
		}
	}
	return nil
}

func validateEvalSuiteDefaults(suite *EvalSuite) error {
	switch {
	case suite.FKF != evalSchemaVersion:
		return fmt.Errorf("fkf must be %d; got %d", evalSchemaVersion, suite.FKF)
	case suite.K < 1 || suite.K > maxEvalK:
		return fmt.Errorf("k is %d; expected 1..%d", suite.K, maxEvalK)
	case suite.Budget != nil && (*suite.Budget < 1 || *suite.Budget > maxEvalBudget):
		return fmt.Errorf("budget is %d; expected 1..%d when set", *suite.Budget, maxEvalBudget)
	case suite.RecallThreshold <= 0 || suite.RecallThreshold > 1:
		return fmt.Errorf("recall_threshold is %g; expected greater than 0 and at most 1", suite.RecallThreshold)
	case len(suite.Queries) == 0:
		return errors.New("queries must contain at least one evaluation")
	}
	if suite.Budget == nil {
		budget := DefaultBudget
		suite.Budget = &budget
	}
	if suite.Delivery == "" {
		suite.Delivery = ContextDeliveryJSON
	}
	if err := validateEvalDelivery(suite.Delivery); err != nil {
		return err
	}
	return nil
}

func validateEvalDelivery(delivery string) error {
	switch delivery {
	case "", ContextDeliveryJSON, ContextDeliveryJSONL, ContextDeliveryText:
		return nil
	default:
		return fmt.Errorf("delivery %q must be json, jsonl, or text", delivery)
	}
}

func validateEvalQuery(query *EvalQuery, index int, names map[string]struct{}) error {
	query.Name = strings.TrimSpace(query.Name)
	query.Question = strings.TrimSpace(query.Question)
	if query.Name == "" || query.Question == "" {
		return fmt.Errorf("queries[%d] needs non-empty name and question", index)
	}
	if _, exists := names[query.Name]; exists {
		return fmt.Errorf("queries[%d].name %q is duplicated", index, query.Name)
	}
	names[query.Name] = struct{}{}
	if err := validateEvalDelivery(query.Delivery); err != nil {
		return fmt.Errorf("queries[%d]: %w", index, err)
	}
	if query.K != nil && (*query.K < 1 || *query.K > maxEvalK) {
		return fmt.Errorf("queries[%d].k is %d; expected 1..%d when set", index, *query.K, maxEvalK)
	}
	if query.Budget != nil && (*query.Budget < 1 || *query.Budget > maxEvalBudget) {
		return fmt.Errorf("queries[%d].budget is %d; expected 1..%d when set", index, *query.Budget, maxEvalBudget)
	}
	if query.ExpectEmpty && (len(query.ExpectedURIs) > 0 || len(query.ForbiddenURIs) > 0) {
		return fmt.Errorf("queries[%d].expect_empty cannot be combined with expected_uris or forbidden_uris", index)
	}
	if !query.ExpectEmpty && len(query.ExpectedURIs) == 0 {
		return fmt.Errorf("queries[%d].expected_uris must contain at least one URI unless expect_empty is true", index)
	}
	expected, err := canonicalEvalURIs(query.ExpectedURIs)
	if err != nil {
		return fmt.Errorf("queries[%d].expected_uris: %w", index, err)
	}
	forbidden, err := canonicalEvalURIs(query.ForbiddenURIs)
	if err != nil {
		return fmt.Errorf("queries[%d].forbidden_uris: %w", index, err)
	}
	for _, uri := range forbidden {
		if slices.Contains(expected, uri) {
			return fmt.Errorf("queries[%d] URI %q is both expected and forbidden", index, uri)
		}
	}
	query.ExpectedURIs, query.ForbiddenURIs = expected, forbidden
	return nil
}

func canonicalEvalURIs(values []string) ([]string, error) {
	canonical := make([]string, 0, len(values))
	seen := make(map[string]struct{}, len(values))
	for _, value := range values {
		uri, err := ParseURI(value)
		if err != nil {
			return nil, err
		}
		rendered := uri.String()
		if _, exists := seen[rendered]; exists {
			return nil, fmt.Errorf("URI %q is duplicated", rendered)
		}
		seen[rendered] = struct{}{}
		canonical = append(canonical, rendered)
	}
	return canonical, nil
}

File: src/fkf/eval.py
"""Strict, offline recall-at-k evaluation over the context delivery contract."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Final

from pydantic import BaseModel, ConfigDict, ValidationError
from pydantic import Field as PydanticField

from fkf.base import Base
from fkf.config import ConfigError, decode_strict_yaml
from fkf.context import (
    CONTEXT_DELIVERY_JSON,
    CONTEXT_DELIVERY_JSONL,
    CONTEXT_DELIVERY_TEXT,
    DEFAULT_BUDGET,
    ContextPack,
    ContextRequest,
    _ContextBatchError,
    _map_contexts,
    render_context_bytes,
)
from fkf.io import read_file_limited
from fkf.jsoncodec import dumps
from fkf.lexical import RANKING_VERSION, LexicalIndexUse
from fkf.output import register_text
from fkf.process import Cancellation
from fkf.query import Window
from fkf.read import read
from fkf.store import MAX_CONFIG_BYTES, MAX_NARRATIVE_BYTES, validate_within_root
from fkf.uri import URIError, parse_uri

EVAL_SCHEMA_VERSION: Final = 1
EVAL_PATH: Final = "checks/queries.yaml"
MAX_EVAL_K: Final = 100
MAX_EVAL_BUDGET: Final = MAX_NARRATIVE_BYTES // 4
_DELIVERIES: Final = frozenset({CONTEXT_DELIVERY_JSON, CONTEXT_DELIVERY_JSONL, CONTEXT_DELIVERY_TEXT})


class _Boundary(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


class _WindowModel(_Boundary):
    since: str = ""
    until: str = ""


class _QueryModel(_Boundary):
    name: str
    question: str
    window: _WindowModel = _WindowModel()
    k: int | None = None
    budget: int | None = None
    delivery: str = ""
    expect_empty: bool = False
    expected_uris: list[str] = PydanticField(default_factory=list)
    forbidden_uris: list[str] = PydanticField(default_factory=list)
    expected_excerpts: dict[str, list[str]] = PydanticField(default_factory=dict)
    expected_reads: dict[str, list[str]] = PydanticField(default_factory=dict)


class _SuiteModel(_Boundary):
    fkf: int
    k: int
    budget: int | None = None
    delivery: str = ""
    recall_threshold: float
    queries: list[_QueryModel]


@dataclass(frozen=True, slots=True)
class EvalExpectedRank:
    uri: str
    rank: int


@dataclass(frozen=True, slots=True)
class EvalQueryResult:
    name: str
    question: str
    window: Window
    budget: int
    delivery: str
    k: int
    expect_empty: bool
    recall: float
    expected: int
    found_expected: int
    expected_ranks: tuple[EvalExpectedRank, ...]
    missing_expected: tuple[str, ...] = field(default=(), metadata={"json": "missing_expected,omitempty"})
    forbidden_found: tuple[str, ...] = field(default=(), metadata={"json": "forbidden_found,omitempty"})
    delivered_uris: tuple[str, ...] = ()
    delivered_bytes: int = 0
    delivered_tokens: int = 0
    index: LexicalIndexUse = field(default_factory=LexicalIndexUse)
    input_digest: str = ""
    ranking_version: int = RANKING_VERSION
    recall_threshold: float = 0.0
    passed: bool = False
    missing_evidence: tuple[str, ...] = field(default=(), metadata={"json": "missing_evidence,omitempty"})


@dataclass(frozen=True, slots=True)
class EvalReport:
    path: str
    k: int
    budget: int
    delivery: str
    recall_threshold: float
    evaluation_time: str
    queries: tuple[EvalQueryResult, ...]
    passed_queries: int
    failed: int = field(metadata={"json": "failed_queries"})
    passed: bool = False


@dataclass(frozen=True, slots=True)
class _EvalQuery:
    name: str
    question: str
    window: Window
    k: int | None
    budget: int | None
    delivery: str
    expect_empty: bool
    expected_uris: tuple[str, ...]
    forbidden_uris: tuple[str, ...]
    expected_excerpts: dict[str, tuple[str, ...]]
    expected_reads: dict[str, tuple[str, ...]]


@dataclass(frozen=True, slots=True)
class _EvalSuite:
    k: int
    budget: int
    delivery: str
    recall_threshold: float
    queries: tuple[_EvalQuery, ...]


def _invalid(message: str, *, cause: BaseException | None = None) -> ConfigError:
    return ConfigError(f"invalid configuration: {message}", cause=cause)


def _format_evaluation_time(value: datetime, *, seconds: bool = False) -> str:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError("evaluation timestamp must include a timezone")
    rendered = value.isoformat(timespec="seconds" if seconds else "auto")
    return rendered.removesuffix("+00:00") + "Z" if rendered.endswith("+00:00") else rendered


def _delivery(value: str, label: str) -> str:
    selected = value or CONTEXT_DELIVERY_JSON
    if selected not in _DELIVERIES:
        raise _invalid(f"{label}delivery {selected!r} must be json, jsonl, or text")
    return selected


def _canonical_uris(values: tuple[str, ...], label: str) -> tuple[str, ...]:
    result: list[str] = []
    seen: set[str] = set()
    for value in values:
        try:
            rendered = str(parse_uri(value))
        except URIError as error:
            raise _invalid(f"{label}: {error}", cause=error) from error
        if rendered in seen:
            raise _invalid(f"{label}: URI {rendered!r} is duplicated")
        seen.add(rendered)
        result.append(rendered)
    return tuple(result)


def _query(value: _QueryModel, index: int, names: set[str]) -> _EvalQuery:
    name = value.name.strip()
    question = value.question.strip()
    if not name or not question:
        raise _invalid(f"{EVAL_PATH}: queries[{index}] needs non-empty name and question")
    if name in names:
        raise _invalid(f"{EVAL_PATH}: queries[{index}].name {name!r} is duplicated")
    names.add(name)
    if value.k is not None and not 1 <= value.k <= MAX_EVAL_K:
        raise _invalid(f"{EVAL_PATH}: queries[{index}].k is {value.k}; expected 1..{MAX_EVAL_K} when set")
    if value.budget is not None and not 1 <= value.budget <= MAX_EVAL_BUDGET:
        raise _invalid(
            f"{EVAL_PATH}: queries[{index}].budget is {value.budget}; expected 1..{MAX_EVAL_BUDGET} when set"
        )
    delivery = _delivery(value.delivery, f"{EVAL_PATH}: queries[{index}]: ") if value.delivery else ""
    if value.expect_empty and (value.expected_uris or value.forbidden_uris):
        raise _invalid(
            f"{EVAL_PATH}: queries[{index}].expect_empty cannot be combined with expected_uris or forbidden_uris"
        )
    if not value.expect_empty and not value.expected_uris:
        raise _invalid(
            f"{EVAL_PATH}: queries[{index}].expected_uris must contain at least one URI unless expect_empty is true"
        )
    expected = _canonical_uris(tuple(value.expected_uris), f"{EVAL_PATH}: queries[{index}].expected_uris")
    forbidden = _canonical_uris(tuple(value.forbidden_uris), f"{EVAL_PATH}: queries[{index}].forbidden_uris")
    overlap = next((uri for uri in forbidden if uri in expected), "")
    if overlap:
        raise _invalid(f"{EVAL_PATH}: queries[{index}] URI {overlap!r} is both expected and forbidden")
    for kind, evidence in (("expected_excerpts", value.expected_excerpts), ("expected_reads", value.expected_reads)):
        for uri, snippets in evidence.items():
            if uri not in expected or not snippets or any(not snippet.strip() for snippet in snippets):
                raise _invalid(
                    f"{EVAL_PATH}: queries[{index}].{kind} requires expected URI keys and non-empty text lists"
                )
    return _EvalQuery(
        name,
        question,
        Window(value.window.since, value.window.until),
        value.k,
        value.budget,
        delivery,
        value.expect_empty,
        expected,
        forbidden,
        {uri: tuple(snippets) for uri, snippets in value.expected_excerpts.items()},
        {uri: tuple(snippets) for uri, snippets in value.expected_reads.items()},
    )


def _load_suite(base: Base) -> _EvalSuite:
    absolute = base.root / "checks" / "queries.yaml"
    validate_within_root(base.root, absolute)
    try:
        data = read_file_limited(absolute, MAX_CONFIG_BYTES)
    except OSError as error:
        if isinstance(error.__cause__, FileNotFoundError):
            raise _invalid(f"{EVAL_PATH} is missing; add the base's retrieval evaluation set", cause=error) from error
        raise
    try:
        raw = decode_strict_yaml(data, Path(EVAL_PATH))
        model = _SuiteModel.model_validate(raw)
    except ConfigError:
        raise
    except ValidationError as error:
        problem = error.errors(include_url=False)[0]
        location = ".".join(str(part) for part in problem["loc"])
        message = "field not found" if problem["type"] == "extra_forbidden" else str(problem["msg"])
        raise _invalid(f"{EVAL_PATH}: {location}: {message}", cause=error) from error
    if model.fkf != EVAL_SCHEMA_VERSION:
        raise _invalid(f"{EVAL_PATH}: fkf must be {EVAL_SCHEMA_VERSION}; got {model.fkf}")
    if not 1 <= model.k <= MAX_EVAL_K:
        raise _invalid(f"{EVAL_PATH}: k is {model.k}; expected 1..{MAX_EVAL_K}")
    budget = model.budget if model.budget is not None else DEFAULT_BUDGET
    if not 1 <= budget <= MAX_EVAL_BUDGET:
        raise _invalid(f"{EVAL_PATH}: budget is {budget}; expected 1..{MAX_EVAL_BUDGET} when set")
    if not 0 < model.recall_threshold <= 1:
        raise _invalid(
            f"{EVAL_PATH}: recall_threshold is {model.recall_threshold:g}; expected greater than 0 and at most 1"
        )
    if not model.queries:
        raise _invalid(f"{EVAL_PATH}: queries must contain at least one evaluation")
    names: set[str] = set()
    queries = tuple(_query(value, index, names) for index, value in enumerate(model.queries))
    return _EvalSuite(model.k, budget, _delivery(model.delivery, f"{EVAL_PATH}: "), model.recall_threshold, queries)


def _context_request(suite: _EvalSuite, query: _EvalQuery, evaluation_time: datetime) -> ContextRequest:
    budget = query.budget if query.budget is not None else suite.budget
    delivery = query.delivery or suite.delivery
    return ContextRequest(
        query.question,
        query.window,
        budget,
        delivery_format=delivery,
        evaluation_time=evaluation_time,
    )


def _evaluate_query(
    base: Base, suite: _EvalSuite, query: _EvalQuery, pack: ContextPack, cancel: Cancellation | None
) -> EvalQueryResult:
    budget = query.budget if query.budget is not None else suite.budget
    k = query.k if query.k is not None else suite.k
    delivery = query.delivery or suite.delivery
    delivered = tuple(item.uri for item in pack.items)
    ranks = {uri: index for index, uri in enumerate(delivered, start=1)}
    expected_ranks = tuple(EvalExpectedRank(uri, ranks.get(uri, 0)) for uri in query.expected_uris)
    missing = tuple(item.uri for item in expected_ranks if item.rank == 0 or item.rank > k)
    found = len(expected_ranks) - len(missing)
    forbidden = tuple(uri for uri in query.forbidden_uris if uri in ranks)
    encoded = render_context_bytes(pack, delivery)
    missing_evidence: list[str] = []
    for kind, evidence in (("excerpt", query.expected_excerpts), ("read", query.expected_reads)):
        for uri, snippets in evidence.items():
            item = next((item for item in pack.items[:k] if item.uri == uri), None)
            content = item.excerpt if item is not None else ""
            if item is not None and kind == "read":
                result = read(base, uri, cancel=cancel)
                content = result.text or (dumps(result.record).decode() if result.record is not None else "")
            if any(snippet.casefold() not in content.casefold() for snippet in snippets):
                missing_evidence.append(f"{kind}: {uri}")
    if query.expect_empty:
        recall = 0.0 if delivered or pack.matched_but_omitted else 1.0
        passed = not delivered and not pack.matched_but_omitted
    else:
        recall = found / len(query.expected_uris)
        passed = recall >= suite.recall_threshold and not forbidden and not missing_evidence
    return EvalQueryResult(
        query.name,
        query.question,
        pack.receipt.window,
        budget,
        delivery,
        k,
        query.expect_empty,
        recall,
        len(query.expected_uris),
        found,
        expected_ranks,
        missing,
        forbidden,
        delivered,
        len(encoded),
        (len(encoded) + 3) // 4,
        pack.receipt.index,
        pack.receipt.input_digest,
        pack.receipt.ranking_version,
        suite.recall_threshold,
        passed,
        tuple(missing_evidence),
    )


def evaluate(base: Base, *, cancel: Cancellation | None = None) -> EvalReport:
    """Run every declared query against stored evidence without writing or executing."""

    suite = _load_suite(base)
    evaluation_time = base.now()
    requests = tuple(_context_request(suite, query, evaluation_time) for query in suite.queries)
    try:
        results = _map_contexts(
            base,
            requests,
            lambda index, pack: _evaluate_query(base, suite, suite.queries[index], pack, cancel),
            cancel=cancel,
        )
    except _ContextBatchError as failure:
        from fkf.errors import CanceledError, InvalidUsageError, OperationalError

        error = failure.error
        query = suite.queries[failure.index]
        if isinstance(error, CanceledError):
            raise error from failure
        if isinstance(error, InvalidUsageError):
            raise InvalidUsageError(f"evaluate {query.name}: {error}", cause=error) from error
        if isinstance(error, OperationalError):
            raise OperationalError(f"evaluate {query.name}: {error}", cause=error) from error
        raise error from failure
    passed_queries = sum(result.passed for result in results)
    failed = len(results) - passed_queries
    return EvalReport(
        EVAL_PATH,
        suite.k,
        suite.budget,
        suite.delivery,
        suite.recall_threshold,
        _format_evaluation_time(evaluation_time),
        tuple(results),
        passed_queries,
        failed,
        failed == 0,
    )


def _text(report: EvalReport) -> str:
    lines: list[str] = []
    for query in report.queries:
        state = "PASS" if query.passed else "FAIL"
        lines.append(
            f"{state} {query.name:<24} recall@{query.k} {query.recall:.3f} "
            f"({query.found_expected}/{query.expected}) · budget {query.budget} · {query.delivery} delivered "
            f"{len(query.delivered_uris)} items, {query.delivered_tokens} tokens, {query.delivered_bytes} bytes"
        )
        if query.expected_ranks:
            rendered = "[" + " ".join(f"{{{item.uri} {item.rank}}}" for item in query.expected_ranks) + "]"
            lines.append(f"  expected ranks: {rendered}")
        if query.missing_expected:
            lines.append(f"  missing: [{' '.join(query.missing_expected)}]")
        if query.forbidden_found:
            lines.append(f"  forbidden: [{' '.join(query.forbidden_found)}]")
        if query.missing_evidence:
            lines.append(f"  missing evidence: [{', '.join(query.missing_evidence)}]")
    lines.append(
        f"{report.passed_queries} passed, {report.failed} failed · threshold {report.recall_threshold:.3f} · "
        f"default k {report.k} · default budget {report.budget} · "
        f"evaluated {_format_evaluation_time(datetime.fromisoformat(report.evaluation_time), seconds=True)} "
        f"· {report.path}"
    )
    return "\n".join(lines)


register_text(EvalReport, _text)


__all__ = [
    "EVAL_PATH",
    "EVAL_SCHEMA_VERSION",
    "MAX_EVAL_BUDGET",
    "MAX_EVAL_K",
    "EvalExpectedRank",
    "EvalQueryResult",
    "EvalReport",
    "evaluate",
]
